### PR TITLE
Fix connection issue in dbt 1.0.0

### LINKED ISSUE #2113: dbt.exceptions.InvalidConnectionException: Runtime Error on 0.9.0
I'm trying to [upgrade my project](https://github.com/gnilrets/dbt-container-skeleton) to dbt 1.0.0 and SQLFluff 0.9.0.  I can `dbt run` just fine, but when I run SQLFluff, I get a connection error.


Here's the full stacktrace of attempting to lint my project using SQLFluff:
```
root@541f43ad6080:/dbt-runner/dbt# sqlfluff lint models
=== [dbt templater] Sorting Nodes...
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
file dim_dates.sql:  11%|██████████████████████▏                                                                                                                                                                                 | 1/9 [00:00<00:05,  1.49it/s]Traceback (most recent call last):
  File "/usr/local/bin/sqlfluff", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/cli/commands.py", line 480, in lint
    result = lnt.lint_paths(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 964, in lint_paths
    self.lint_path(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 916, in lint_path
    for i, linted_file in enumerate(runner.run(fnames, fix), start=1):
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter/runner.py", line 101, in run
    for fname, partial in self.iter_partials(fnames, fix=fix):
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter/runner.py", line 54, in iter_partials
    for fname, rendered in self.iter_rendered(fnames):
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter/runner.py", line 43, in iter_rendered
    yield fname, self.linter.render_file(fname, self.config)
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 642, in render_file
    return self.render_string(raw_file, fname, config, encoding)
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 619, in render_string
    templated_file, templater_violations = self.templater.process(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff_templater_dbt/templater.py", line 330, in process
    processed_result = self._unsafe_process(fname_absolute_path, in_str, config)
  File "/usr/local/lib/python3.8/site-packages/sqlfluff_templater_dbt/templater.py", line 449, in _unsafe_process
    node = self.dbt_compiler.compile_node(
  File "/usr/local/lib/python3.8/site-packages/dbt/compilation.py", line 550, in compile_node
    node = self._compile_node(node, manifest, extra_context)
  File "/usr/local/lib/python3.8/site-packages/dbt/compilation.py", line 388, in _compile_node
    compiled_node.compiled_sql = jinja.get_rendered(
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 597, in get_rendered
    return render_template(template, ctx, node)
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 548, in render_template
    return template.render(ctx)
  File "/usr/local/lib/python3.8/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/usr/local/lib/python3.8/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/usr/local/lib/python3.8/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 2, in top-level template code
  File "/usr/local/lib/python3.8/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 332, in __call__
    return self.call_macro(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 259, in call_macro
    return macro(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "<template>", line 2, in template
  File "/usr/local/lib/python3.8/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 332, in __call__
    return self.call_macro(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 259, in call_macro
    return macro(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "<template>", line 16, in template
  File "/usr/local/lib/python3.8/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 332, in __call__
    return self.call_macro(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 259, in call_macro
    return macro(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "<template>", line 2, in template
  File "/usr/local/lib/python3.8/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 332, in __call__
    return self.call_macro(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 259, in call_macro
    return macro(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "<template>", line 2, in template
  File "/usr/local/lib/python3.8/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 332, in __call__
    return self.call_macro(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 259, in call_macro
    return macro(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "<template>", line 10, in template
  File "/usr/local/lib/python3.8/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/dbt/adapters/base/impl.py", line 225, in execute
    return self.connections.execute(
  File "/usr/local/lib/python3.8/site-packages/dbt/adapters/sql/connections.py", line 127, in execute
    _, cursor = self.add_query(sql, auto_begin)
  File "/usr/local/lib/python3.8/site-packages/dbt/adapters/sql/connections.py", line 59, in add_query
    connection = self.get_thread_connection()
  File "/usr/local/lib/python3.8/site-packages/dbt/adapters/base/connections.py", line 69, in get_thread_connection
    raise dbt.exceptions.InvalidConnectionException(
dbt.exceptions.InvalidConnectionException: Runtime Error
  connection never acquired for thread (82, 140226265859904), have []
```


## Version
```
root@541f43ad6080:/dbt-runner/dbt# sqlfluff --version
sqlfluff, version 0.9.0
```

```
root@3c162864a926:/dbt-runner/dbt# python --version
Python 3.8.12
```

output of dbt debug:
```
root@541f43ad6080:/dbt-runner/dbt# dbt debug                                                                                   
21:50:16  Running with dbt=1.0.0                     
dbt version: 1.0.0                       
python version: 3.8.12                                                                                                         
python path: /usr/local/bin/python               
os info: Linux-5.10.47-linuxkit-x86_64-with-glibc2.2.5                                                                         
Using profiles.yml file at /dbt-runner/dbt/profiles.yml
Using dbt_project.yml file at /dbt-runner/dbt/dbt_project.yml                                                                  
                                                               
Configuration:                                                                                                                 
  profiles.yml file [OK found and valid]          
  dbt_project.yml file [OK found and valid]                                                                                    
                                                                                                                               
Required dependencies:                                                                                                         
 - git [OK found]                                                                                                              
                                                                                                                               
Connection:                                                                                                                    
  host: host.docker.internal                                                                                                   
  port: 5432                                                                                                                   
  user: sterling.paramore                                                                                                      
  database: blert                                             
  schema: wh                                                                                                                   
  search_path: None                                   
  keepalives_idle: 0                                                                                                           
  sslmode: None                                             
  Connection test: [OK connection ok]                                                                                          
                                                                                                                               
All checks passed!
```


## Configuration
```
[sqlfluff]
verbose = 0
nocolor = False
dialect = ansi
templater = dbt
rules = None
exclude_rules = L009
recurse = 0
output_line_length = 80
runaway_limit = 10
ignore_templated_areas = True
# Comma separated list of file extensions to lint.

# NB: This config will only apply in the root folder.
sql_file_exts = .sql,.sql.j2,.dml,.ddl

[sqlfluff:indentation]
indented_joins = False
template_blocks_indent = True

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}dbt_{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable) %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 110
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config.
[sqlfluff:rules:L003]
lint_templated_tokens = True

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = consistent

[sqlfluff:rules:L014]  # Unquoted identifiers
extended_capitalisation_policy = consistent

[sqlfluff:rules:L016]
ignore_comment_lines = False

[sqlfluff:rules:L029]  # Keyword identifiers
unquoted_identifiers_policy = aliases

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = consistent

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = consistent

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = join

[sqlfluff:rules:L047]  # Consistent syntax to count all rows
prefer_count_1 = False
```

### PR DESCRIPTION
**Pull request by NiallRees**
Resolves #2113 

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: .github/workflows/ci-tests.yml
Comment (NiallRees) [inline .github/workflows/ci-tests.yml:114]: Adding a Postgres database to run the dbt templater against

File: tox.ini
Comment (NiallRees) [inline tox.ini:48]: This step has to run in separate invocation because dbt caches the connection to the database, and doesn't load the deliberately broken profile.

Comment (NiallRees) [reply tox.ini:48]: There must be a way to reload the database connection in the same invocation, but I can't work out how!

File: plugins/sqlfluff-templater-dbt/test/templater_test.py
Comment (NiallRees) [inline plugins/sqlfluff-templater-dbt/test/templater_test.py:344]: What I thought was just a change to dbt 1.0.0 turned out to be a demonstration of the error that this PR solves

File: tox.ini
Comment (jpy-git) [reply tox.ini:48]: @NiallRees I actually have a slightly different suggestion here.

All this test really needs to do is validate how sqlfluff handles the exception when a connection error is returned by dbt. Rather than trying to use a different connection and having to split this out, we can just mock/monkey patch the return of the relevant function in the code to raise the connection error in this test. That way we can test how sqlfluff responds without having to actually trigger a real connection error 😄

[pytest-mock](https://pypi.org/project/pytest-mock/) is the package you want if you're interested.
I use it for things like testing how my code handles unlikely connection errors to databases/APIs 👍

**Review by jpy-git — COMMENTED** [review]
Some general thoughts (bear in mind I'm not a dbt pro) 😄

File: tox.ini
Comment (barrywhart) [reply tox.ini:48]: You can also use the built-in Python module `unittest.mock`. It provides a decorator, so it's pretty easy to use. https://otrabalhador.github.io/python-testing-by-examples/docs/en/mocking/essentials/patch.html

File: CONTRIBUTING.md
Comment (barrywhart) [reply CONTRIBUTING.md:173]: +1 for providing a Docker Compose configuration for running the dbt tests. I actually have some bits of this on my local machine already, because prior to getting dbt 1.0 working, I couldn't install dbt (older versions) at all.

If we want to create a separate issue for setting up Docker Compose, I can potentially take on part of that work, or at least share what I have done so far.

Comment (jpy-git) [reply CONTRIBUTING.md:173]: Amazing, could you create the issue and share what you've got already? Can just take the motivation/description from my comment above 😄

Comment (barrywhart) [reply CONTRIBUTING.md:173]: Done! #2248

**Comment by hugohjerten** [comment]
Appreciating the effort here! 🙌  Just tried this revision out to see if it fixes the issue for me, however I'm now experiencing a new error:
```sh
ModuleNotFoundError: No module named 'cached_property'
```
`cached_property` doesn't seem to be available 🤷 (Don't know if this is due to a change in this PR, or other PR's).

**Comment by jpy-git** [comment]
> Appreciating the effort here! 🙌 Just tried this revision out to see if it fixes the issue for me, however I'm now experiencing a new error:
> 
> ```shell
> ModuleNotFoundError: No module named 'cached_property'
> ```
> 
> `cached_property` doesn't seem to be available 🤷 (Don't know if this is due to a change in this PR, or other PR's).

@hugohjerten We updated the `cached_property` dependency in #2212. We have verified the this works fine in the CI tests for all versions (you can see it works fine for this PR too).

My suspicion is that you've checked out this branch but not reinstalled it?
Did you do that pip install method I suggested in the linked issue?
If you're using this branch locally and haven't account for recent dependency updates you should reinstall your venv `tox -e py -r --devenv .venv && source .venv/bin/activate`

**Comment by hugohjerten** [comment]
@jpy-git I use poetry, and removed my `venv`, and even my `poetry.lock` file to get a proper clean slate, and still encountered this issue 🤷 . Maybe I discovered a case that the CI tests don't conver? 🙂

**Comment by jpy-git** [comment]
> @jpy-git I use poetry, and removed my `venv`, and even my `poetry.lock` file to get a proper clean slate, and still encountered this issue 🤷 . Maybe I discovered a case that the CI tests don't cover? 🙂

Can you open an issue with a minimum reproducible example of the commands/config that you use to setup your virtual environment please? Seems like a poetry specific thing (potentially only in the case you install from a git branch) so let's look at it separately.

File: tox.ini
Comment (NiallRees) [reply tox.ini:48]: As we have a database connection I thought we might as well break it for real, but perhaps mocking is the best option.

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py
Comment (NiallRees) [reply plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:455]: > > @jpy-git I use poetry, and removed my `venv`, and even my `poetry.lock` file to get a proper clean slate, and still encountered this issue 🤷 . Maybe I discovered a case that the CI tests don't cover? 🙂
> 
> Can you open an issue with a minimum reproducible example of the commands/config that you use to setup your virtual environment please? Seems like a poetry specific thing (potentially only in the case you install from a git branch) so let's look at it separately.

@hugohjerten you'll need to run:
```
tox -e dbt100-py39 --devenv .venv
source .venv/bin/activate
```
And then install any adapters you need, e.g.:
```
pip install dbt-snowfalke
```

**Comment by hugohjerten** [comment]
@jpy-git  False alarm. This was caused by `sqlfluff-templater-dbt` not being updated to the same revision also. Sorry! :)

**Comment by jpy-git** [comment]
> @jpy-git False alarm. This was caused by `sqlfluff-templater-dbt` not being updated to the same revision also. Sorry! :)

@hugohjerten ahh good spot! I always forget that they're separate packages too! let us know if the branch helps fix the connection issue 😄

**Comment by hugohjerten** [comment]
✅  QA'd! 😸 I managed to test if this fixes the bug by installing the revision with the following poetry syntax in `pyproject.toml`:
```toml
[tool.poetry.dependencies]
python = "^3.9"
dbt-core = "^1.0.0"
dbt-bigquery = "^1.0.0"

[tool.poetry.dev-dependencies]
sqlfluff = {git = "https://github.com/sqlfluff/sqlfluff", rev = "nw/fix_connection_dbt_1.0.0"}
jinja2-simple-tags = "^0.3.1"
sqlfluff-templater-dbt = {path = ".venv/src/sqlfluff/plugins/sqlfluff-templater-dbt", develop = true}

[build-system]
requires = ["poetry-core>=1.1.0"]
build-backend = "poetry.core.masonry.api"

[tool.sqlfluff.core]
templater = "dbt"
dialect = "bigquery"
```

FYI @jpy-git by explicitly installing the extra dependency (`jinja2-simple-tags`) introduced in `sqlfluff-templater-dbt`, I was able to install the `sqlfluff-templater-dbt` in editable mode in poetry. :)

**Comment by jpy-git** [comment]
> ✅ QA'd! 😸

Yay thanks @hugohjerten! 👏

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py
Comment (NiallRees) [reply plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:455]: > Nice job finding this! 😄
> 
> Part of me wonders given how many people are being impacted by this issue in #2113 and that the change itself is relatively small (and presumably you've verified it locally), we could potentially push just this bit through in time for the next release. 😈 (N.B. I would only advocate this in the event that we think the setup for the live postgres db connection testing is going to take too long to fit this into the next release 👼 )

@jpy-git - unfortunately the new code I've added here now means that we have to have a database connection for the dbt templater to run in dbt >= 1.0.0. Happy to get this merged in the easiest way possible, but we do at the minimum need a database connection for the tests to run.

Comment (barrywhart) [reply plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:455]: Would it be useful to try and find a way to make this optional? I'm thinking if the user's dbt models don't use any dbt features that require a database connection, it may be too onerous if SQLFluff requires it, especially for CI/CD environments.

Is there a way we could try and connect, but continue gracefully if we can't?

Comment (jpy-git) [reply plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:455]: (maybe even mock out the connection line in a fixture for the dbt tests)

Comment (barrywhart) [reply plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:455]: Oh, duh, my previous message was "off": I was thinking of dbt **users** always requiring a database, not the **tests**. Mocking does seem like a possibility.

Question: How valuable is it for the tests to use an actual database, versus something like mocks? In other words, if we mock the database connection, how much risk do we see that this may miss a future issue that having a real database would detect? This may depend on the kind of tests we have now, or may add in the future. (I could imagine more and more tests needing this over time, depending on what sort of issues our users report.)

**Comment by barrywhart** [comment]
@jpy-git, @NiallRees: I have a draft PR for the Docker Compose environment: #2254.

It installs the dbt templater and dependencies, and it creates a second Postgres container. For now, I have mainly focused on verifying the core SQLFluff tests work correctly. The dbt bits are installed, but I haven't tried running those tests. Should I try and do that now? I'm unsure whether to do that now, or wait until **this** PR is merged.

TBH, I think we need to resolve the discussion about whether to mock Postgres first -- that could help break us out of this slight "chicken-egg" situation we're in. What do y'all think?

**Comment by NiallRees** [comment]
> @jpy-git, @NiallRees: I have a draft PR for the Docker Compose environment: #2254.
> 
> It installs the dbt templater and dependencies, and it creates a second Postgres container. For now, I have mainly focused on verifying the core SQLFluff tests work correctly. The dbt bits are installed, but I haven't tried running those tests. Should I try and do that now? I'm unsure whether to do that now, or wait until **this** PR is merged.
> 
> TBH, I think we need to resolve the discussion about whether to mock Postgres first -- that could help break us out of this slight "chicken-egg" situation we're in. What do y'all think?

Unfortunately I found that dbt 1.0.0 now absolutely requires a database connection to be able to compile the models. The error that this PR is actually solving is making sure those connections exist, otherwise it errors out when attempting to compile the nodes.

File: tox.ini
Comment (barrywhart) [reply tox.ini:48]: I'm trying to run this test from the PR branch, and I'm getting an error about a missing file:
```
>       os.rename(src_fpath, target_fpath)
E       FileNotFoundError: [Errno 2] No such file or directory: 'plugins/sqlfluff-templater-dbt/test/fixtures/dbt/error_models/exception_connect_database.sql' -> '/Users/bhart/dev/sqlfluff/plugins/sqlfluff-templater-dbt/test/fixtures/dbt/dbt_project/models/my_new_project/exception_connect_database.sql'

plugins/sqlfluff-templater-dbt/test/templater_test.py:339: FileNotFoundError

 plugins/sqlfluff-templater-dbt/test/templater_test.py ⨯                                                                                                                                 100% ██████████
======================================================================================= short test summary info ========================================================================================
FAILED plugins/sqlfluff-templater-dbt/test/templater_test.py::test__templater_dbt_handle_database_connection_failure - FileNotFoundError: [Errno 2] No such file or directory: 'plugins/sqlfluff-temp...
```

@NiallRees: Do you have a file `plugins/sqlfluff-templater-dbt/test/fixtures/dbt/dbt_project/models/my_new_project/exception_connect_database.sql` that needs committing?

Comment (barrywhart) [reply tox.ini:48]: Disregard my question about a missing file. My local clone was messed up somehow.

File: plugins/sqlfluff-templater-dbt/test/fixtures/dbt/profiles_yml/profiles.yml
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/test/fixtures/dbt/profiles_yml/profiles.yml:6]: @jpy-git: Note for later: In the Docker Compose environment, this will need to be `postgres` instead. We should automate this later, but let's not worry about it right now, with two separate PRs in flight (dbt fix, Docker Compose). 🤯

**Comment by barrywhart** [comment]
@NiallRees: Understood (re: database connection being required now). Do you need help from @jpy-git or me getting this PR finished? If so, my Docker Compose PR (#2254) is ready for review and potential merge.

If we can get that merged and get this PR updated, that will make it easier for us to assist. (Note there is a merge conflict at the moment -- it's pretty easy to resolve, though, just a one line conflict in `tox.ini`.)

Let us know if you would like help, and if so, what area(s).

**Comment by NiallRees** [comment]
> @NiallRees: Understood (re: database connection being required now). Do you need help from @jpy-git or me getting this PR finished? If so, my Docker Compose PR (#2254) is ready for review and potential merge.
> 
> If we can get that merged and get this PR updated, that will make it easier for us to assist. (Note there is a merge conflict at the moment -- it's pretty easy to resolve, though, just a one line conflict in `tox.ini`.)
> 
> Let us know if you would like help, and if so, what area(s).

Please feel free to move this forward in any way that your time and interest permits! The outstanding pieces are in the PR description.

**Comment by barrywhart** [comment]
@NiallRees, @jpy-git: Ok, I updated `test__templater_dbt_handle_database_connection_failure` to use `mock` to simulate a connection error. With this change, the test passes even when it runs after other tests that use a real database connection.

I don't know anything about the Windows tests in CI, so I'll leave that for someone else.

**Comment by jpy-git** [comment]
Nice I'll take a look at the windows thing in a bit 😊

**Comment by tunetheweb** [comment]
Hmmmm..... this may not be possible: https://github.community/t/how-to-run-services-on-windows-or-macos/16565

**Comment by kochalex** [comment]
> Hmmmm..... this may not be possible: https://github.community/t/how-to-run-services-on-windows-or-macos/16565

Postgres is shown as  available in the default github runner windows image - it just has to be started up - but this would be running int on the same host, not in an isolated container.

**Comment by tunetheweb** [comment]
> > Hmmmm..... this may not be possible: https://github.community/t/how-to-run-services-on-windows-or-macos/16565
> 
> Postgres is shown as available in the default github runner windows image - it just has to be started up - but this would be running int on the same host, not in an isolated container.

I cheated and instead disabled those tests for windows run in https://github.com/sqlfluff/sqlfluff/pull/2230/commits/d75c82341e244cc85a85b94ec408b612e4a29597

**Comment by jpy-git** [comment]
> I cheated and instead disabled those tests for windows run in [d75c823](https://github.com/sqlfluff/sqlfluff/commit/d75c82341e244cc85a85b94ec408b612e4a29597)

@tunetheweb I haven't had a chance to look into the Postgres Windows CI setup yet, but in theory I'm happy with disabling them for now so as not to block this and I'll investigate and follow up with a PR to re-add them if possible.

Reviewing now 👀

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:57]: What's this for?

File: tox.ini
Comment (barrywhart) [reply tox.ini:48]: We can now combine the two lines above, I think? (Now that the snow connection mock failure is in place.)

**Review by barrywhart — CHANGES_REQUESTED** [review]

**Comment by jpy-git** [comment]
@barrywhart FYI am making tweaks here: #2257

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py
Comment (jpy-git) [reply plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:57]: @barrywhart let's try removing this in my PR and see what happens 😛

Comment (jpy-git) [reply plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:57]: image
assume it is an option dbt takes, but it's false by default so can probs just remove

File: CONTRIBUTING.md
Comment (barrywhart) [reply CONTRIBUTING.md:173]: Done!

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py
Comment (jpy-git) [reply plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:57]: Ooh okay it threw an error when removed let's leave it 🤣 
image

File: CONTRIBUTING.md
Comment (jpy-git) [reply CONTRIBUTING.md:173]: thanks :)

**Comment by barrywhart** [comment]
🍿

File: CONTRIBUTING.md
Comment (jpy-git) [reply CONTRIBUTING.md:173]: @barrywhart actually looks like you haven't pushed the doc changes?

Comment (barrywhart) [reply CONTRIBUTING.md:173]: Oh yes, I pushed them just now. I usually push to my fork, but with this PR, I need to push to the main SQLFluff repo.

**Review by barrywhart — APPROVED** [review]

**Review by jpy-git — APPROVED** [review]
@NiallRees Happy with this! LGTM 🚀 ✨

**Comment by jpy-git** [comment]
@barrywhart you can do the honours when it's ready 😄

**Comment by NiallRees** [comment]
Go team! 🚀

**Pull request merged**
